### PR TITLE
Add u-boot support

### PR DESCRIPTION
**Pull request by ZubairLK**
This patch adds u-boot in the boot chain in preparation for other more exciting stuff.

Testing: hostapp-update + boot tested and on pi3b+ and pi1b (with #205 applied).

### CODE REVIEW

**Comment by agherzan** [comment]
Add to whitelist

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot/0004-rpi.h-Add-bootcmd.patch [hunk 1]
```diff
@@ -0,0 +1,30 @@
+From 05952034ad5acc9b489e89e8ce6d4d28a5a82ce0 Mon Sep 17 00:00:00 2001
+From: Zubair Lutfullah Kakakhel <zubair@resin.io>
+Date: Tue, 12 Jun 2018 18:32:57 +0100
+Subject: [PATCH 4/5] rpi.h: Add bootcmd
+
+Add a default boot command.
+
+Signed-off-by: Zubair Lutfullah Kakakhel <zubair@resin.io>
```
Comment (agherzan) [inline layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot/0004-rpi.h-Add-bootcmd.patch]: Upstream-status missed.

File: layers/meta-resin-raspberrypi/conf/samples/local.conf.sample [hunk 2]
```diff
@@ -16,6 +16,9 @@ GPU_MEM = "16"
 # RPI BSP uses uncompressed kernel images by default
 KERNEL_IMAGETYPE="zImage"
 
+# RPI Use u-boot. Set this to 1 to use u-boot
```
Comment (agherzan) [inline layers/meta-resin-raspberrypi/conf/samples/local.conf.sample]: I'm not sure why you add it here and afterwards change it to 1 in the next commit. I would just set it to 1 in the next commit.

File: layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend [hunk 3]
```diff
@@ -8,15 +8,25 @@ SDIMG_KERNELIMAGE_raspberrypi3-64 ?= "kernel8.img"
 # Customize resinos-img
```
Comment (agherzan) [inline layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend]: I would split this changes in another commit - for resin-image.

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot/0005-rpi-Add-autoboot-configuration-in-defconfigs.patch [hunk 4]
```diff
@@ -0,0 +1,97 @@
+From 706b40d9326a3dcb82d50370324a5921ca3e24ec Mon Sep 17 00:00:00 2001
+From: Zubair Lutfullah Kakakhel <zubair@resin.io>
+Date: Thu, 14 Jun 2018 12:54:14 +0100
+Subject: [PATCH 5/5] rpi: Add autoboot configuration in defconfigs
```
Comment (agherzan) [inline layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot/0005-rpi-Add-autoboot-configuration-in-defconfigs.patch]: Why is autoboot needed for the switch on uboot? Sounds like you are addressing a different issue which should reside in another PR.

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend [hunk 5]
```diff
@@ -0,0 +1,18 @@
+RESIN_EXTERNAL_MMC = "1"
+RESIN_INTERNAL_MMC = "0"
+UBOOT_KCONFIG_SUPPORT = "1"
+inherit resin-u-boot
+
+DEPENDS += "coreutils-native"
```
Comment (agherzan) [inline layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend]: Why is this dependency needed?

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+FILESEXTRAPATHS_prepend := "${THISDIR}/${PN}:"
+
+
+SRC_URI += "file://uEnv.txt_internal \
```
Comment (agherzan) [inline layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend]: As far as I remember this was only needed for flasher types. It shouldn't be needed for any raspberrypi ones.

File: layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend [hunk 7]
```diff
@@ -8,15 +8,25 @@ SDIMG_KERNELIMAGE_raspberrypi3-64 ?= "kernel8.img"
...
 RESIN_BOOT_PARTITION_FILES_rpi = " \
-    ${KERNEL_IMAGETYPE}${KERNEL_INITRAMFS}-${MACHINE}.bin:/${SDIMG_KERNELIMAGE} \
     bcm2835-bootfiles:/ \
     "
 
 python overlay_dtbs_handler () {
     # Add all the dtb files programatically
     if d.getVar('SOC_FAMILY', True) == 'rpi':
         kernel_imagetype = d.getVar('KERNEL_IMAGETYPE', True)
+        sdimg_kernelimage = d.getVar('SDIMG_KERNELIMAGE', True)
```
Comment (agherzan) [inline layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend]: This needs some description in the commit. I'm not sure I understand what is meant to happen here. As well, we can assume that we use uboot - we will not switch around.

File: layers/meta-resin-raspberrypi/conf/samples/local.conf.sample (hunk 2 above)
Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/conf/samples/local.conf.sample]: That was a sneaky way of letting jenkins build test without u-boot and then with u-boot

File: layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend (hunk 3 above)
Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend]: yea splitting would be wiser

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot/0005-rpi-Add-autoboot-configuration-in-defconfigs.patch (hunk 4 above)
Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot/0005-rpi-Add-autoboot-configuration-in-defconfigs.patch]: Autoboot isn't needed for the switch to u-boot. but this shortens the boot time as the default is something like a 2 second delay in u-boot so I had to mess with the defconfigs anyways.

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend (hunk 5 above)
Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend]: u-boot needed this to build. Which is quite strange that this isn't already in one of the inherit/included files.
I think one of the bash scripts required something in coreutils-native.
I'll double check this

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend (hunk 6 above)
Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend]: I see. will remove it

File: layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend (hunk 7 above)
Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend]: One-way move to u-boot makes things much simpler here. So I'll update the patch for that.

File: layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend [hunk 8]
```diff
@@ -8,7 +8,9 @@ SDIMG_KERNELIMAGE_raspberrypi3-64 ?= "kernel8.img"
 # Customize resinos-img
 RESIN_IMAGE_BOOTLOADER_rpi = "bcm2835-bootfiles"
 RESIN_BOOT_PARTITION_FILES_rpi = " \
-    ${KERNEL_IMAGETYPE}${KERNEL_INITRAMFS}-${MACHINE}.bin:/${SDIMG_KERNELIMAGE} \
+		u-boot.bin:/${SDIMG_KERNELIMAGE} \
+		boot.scr:/boot.scr \
+		${KERNEL_IMAGETYPE}${KERNEL_INITRAMFS}-${MACHINE}.bin:/zImage \
```
Comment (agherzan) [inline layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend]: Shouldn't zImage be SDIMG_KERNELIMAGE?

Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend]: `SDIMG_KERNELIMAGE` is actually `kernel.img` `kernel7.img` or `kernel8.img`

But I guess that zImage should be `KERNEL_IMAGETYPE` which is zImage in the local.conf.sample file

Comment (agherzan) [reply layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend]: Yup. I actually meant KERNEL_IMAGETYPE.

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend [hunk 9]
```diff
@@ -0,0 +1,8 @@
+RESIN_EXTERNAL_MMC = "1"
+RESIN_INTERNAL_MMC = "0"
```
[the remark this answers is not in the document]
Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend]: Copied them from the TI u-boot recipe as they looked important for flasher/non-flasher images.
@agherzan any idea where they are used?
I can't find any mention other than inside the bbappend recipes in resin-beaglebone

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot/0007-Fix-misaligned-buffer-in-env_fat_save.patch [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+message in file_fat_write:
+
+  Saving Environment to FAT... writing uboot.env
+  FAT: Misaligned buffer address (9df1c8e0)
+  OK
+
+Upstream-Status: Backport from v2018.03
+
+Signed-off-by: Alex Kiernan <alex.kiernan@gmail.com>
+(cherry picked from commit cda87ec5587261cc813b404f713b102afede0800)
```
[the remark this answers is not in the document]
Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot/0007-Fix-misaligned-buffer-in-env_fat_save.patch]: I followed the wiki where the example had upstream-status above the signed-off-by line..

https://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines#Patch_Header_Recommendations:_Upstream-Status

And I didn't add my sign-off as my sign-off exists in the commit that adds this patch. I did nothing to the patch itself and simply cherry-picked it from upstream.

File: layers/meta-resin-raspberrypi/recipes-support/hostapp-update-hooks/hostapp-update-hooks.bbappend [hunk 11]
```diff
@@ -1,3 +1,3 @@
 FILESEXTRAPATHS_prepend := "${THISDIR}/files:"
 
-HOSTAPP_HOOKS += " 99-rpi-bootloader"
+HOSTAPP_HOOKS += " 99-resin-uboot 99-rpi-bootloader"
```
[the remark this answers is not in the document]
Comment (ZubairLK) [reply layers/meta-resin-raspberrypi/recipes-support/hostapp-update-hooks/hostapp-update-hooks.bbappend]: Yes both hooks are used.
99-resin-uboot updates resinOS_uEnv.txt which we need as our env_resin.h file imports the rootpart from there.

99-rpi-bootloader hook is needed as it updates cmdline.txt. We need to keep cmdline.txt up to date as well. The rpi firmware bootloader is not replaced. That firmware still runs and parses cmdline.txt and populates the fdt /chosen/bootargs node which we still need.
Previously the firmware would load the kernel and boot it.
Now the firmware loads u-boot and u-boot loads the kernel

File: layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend [hunk 12]
```diff
@@ -0,0 +1,6 @@
+UBOOT_KCONFIG_SUPPORT = "1"
```
Comment (agherzan) [inline layers/meta-resin-raspberrypi/recipes-bsp/u-boot/u-boot_%.bbappend]: Nit: Usually I use inherit + variable so you can force with the variable whatever you get from the class. In this case, it's fine because the class uses weak assignments.

File: layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend (hunk 3 above)
Comment (agherzan) [reply layers/meta-resin-raspberrypi/recipes-core/images/resin-image.bbappend]: I would tweak a little bit the commit log: "/mnt/boot" is a path on root. Which happens to be a mountpoint to boot partition but it's confusing. I would say that that it is placed in the root of the boot partition. Whatever root uses to mount that, it's another story.

File: layers/meta-resin-raspberrypi/recipes-support/hostapp-update-hooks/hostapp-update-hooks.bbappend (hunk 11 above)
[the remark this answers is not in the document]
Comment (agherzan) [reply layers/meta-resin-raspberrypi/recipes-support/hostapp-update-hooks/hostapp-update-hooks.bbappend]: cmdline.txt root argument is not needed anymore. And if so, we can remove the `99-rpi-bootloader` hook completely.
